### ISSUE TITLE
gsutil does not take into account `no_proxy` or `NO_PROXY` environment variable

### ISSUE THREAD

**Issue by sonots** [GoogleCloudPlatform/gsutil#435]
Version: google-cloud-sdk-135.0.0-linux-x86_64

With environment variables like:

```
$ env | grep -i proxy
NO_PROXY=127.0.0.1,localhost,www.googleapis.com,.google.com
no_proxy=127.0.0.1,localhost,www.googleapis.com,.google.com
http_proxy=http://myproxy:8080
HTTP_PROXY=http://myproxy:8080
https_proxy=https://myproxy:8080
HTTPS_PROXY=https://myproxy:8080
```

gsutil access to `www.googleapis.com` and `spreadsheets.google.com` via myproxy.

```
$ gsutil -D cat gs://xxxxx
...
connect: (www.googleapis.com, 443)
proxy: ('myproxy', 8080, True, None, None)
...
```

Because of NO_PROXY configuration, it should not access via myproxy.

**Comment by civitaspo** [comment]
Version: google-cloud-sdk-155.0.0-linux-x86_64

`gsutil` has the same behavior as above.

```
$ env | grep -i proxy
NO_PROXY=127.0.0.1,localhost,www.googleapis.com,.google.com
no_proxy=127.0.0.1,localhost,www.googleapis.com,.google.com
http_proxy=http://myproxy:8080
HTTP_PROXY=http://myproxy:8080
https_proxy=https://myproxy:8080
HTTPS_PROXY=https://myproxy:8080
```

```
$ gsutil -D ls
...
gsutil version: 4.26
...
boto version: 2.46.1
...
connect: (www.googleapis.com, 443)
proxy: ('myproxy', 8080, True, None, None)
...
```

**Comment by houglum** [comment]
Thanks for the report, nice find!

We use the `proxy_info_from_url()` method from httplib2, and it looks like it doesn't allow checking the env vars for NO_PROXY/no_PROXY/no_proxy (in order of precedence, [as mentioned here](https://www.gnu.org/software/emacs/manual/html_node/url/Proxies.html)), so the `ProxyInfo` object's `bypass_hosts` attribute never gets populated with the list of domains.  Thus, all subsequent checks for whether or not a request should bypass the proxy return False.

**Comment by houglum** [comment]
My PR got merged into httplib2, so we'll just need to wait for them to release a new version that contains the fix, then we can add it to the requirements in our `setup.py`.

**Comment by houglum** [comment]
Update: With the release of v4.28, we updated to httplib2 v0.10.3... but that was released back in February, and my PR was merged later (in June).  I'm leaving this open, as we're still waiting on the next version of httplib2 to come out so we can mark it as a dependency.

**Comment by houglum** [comment]
I left this sitting on the backburner for a while, apologies! httplib2 has done quite a few releases since my last update; I'll need to pull in the new version into our submodules and update the required version in setup.py so that we can use the upstream fix.

**Comment by houglum** [comment]
Just bumped the httplib2 version. This will be included in the next release of gsutil (v4.31).
